client/src/services/api.js

```diff
@@ -1,9 +1,67 @@
 import axios from 'axios'
 
-let api = axios.create({ baseURL: '/api' })
+let api
 
-function exampleAPICall() {
-  return api.get('/').then((res) => res.data)
+function initApi(token) {
+  if (!token) {
+    //init api without token
+    api = axios.create({ baseURL: '/api' })
+  } else {
+    // if token exists then init authorized api
+    api = axios.create({
+      baseURL: '/api',
+      headers: { authorization: `Bearer ${token}` }
+    })
+  }
 }
 
-export { exampleAPICall }
+// collection of authentication methods
+const authApi = {
+  async signup(formData) {
+    let { data } = await api.post('/auth/signup', formData)
+
+    // get user info to return it back and authorize api
+    if (!data.error) {
+      initApi(data.token)
+      data = (await api.get('/auth/getLoggedUser')).data
+    }
+
+    return data
+  },
+
+  async signin(formData) {
+    let { data } = await api.post('/auth/signin', formData)
+
+    // get user info to return it back and authorize api
+    if (!data.error) {
+      initApi(data.token)
+      data = (await api.get('/auth/getLoggedUser')).data
+    }
+
+    return data
+  },
+
+  async getCachedUser() {
+    try {
+      let { data, status } = await api.get('/auth/getCachedToken', {
+        withCredentials: true
+      })
+
+      // status 204 means there is no cookie, meaning the user is not logged in
+      if (status === 204) {
+        return Promise.resolve({ user: null })
+      }
+
+      initApi(data.token)
+      data = (await api.get('/auth/getLoggedUser')).data
+
+      return data
+    } catch (error) {
+      if (error.response.status === 401) {
+        return Promise.resolve({ user: null })
+      }
+    }
```

Comment:
Why you using try & catch here unlike previous functions?

Reply:
We are using try catch here following the idea that authorization errors should be the responsibility of the api service. Token validation is one of them for example, as it's not a direct result of user input.
This is a good example of `let` or `const` not being the perfect tool as far as code clarity goes, since it's not clear what the try/catch is trying to do here. I think it would have been a better idea to write something like
```
try {
  var { data, status } = await api.get('/auth/getCachedToken', {
    withCredentials: true
  })

  if (status === 204) // ....
} catch (error) {
  // ....
} 

// extract user authentication and api authorization outside the try catch
initApi(data.token)
data = (await api.get('/auth/getLoggedUser')).data
return data
```
to communicate the intent of specifically catching token validation error